File: benchmark/2026-09-17-7677581(latest)/llm-practical/trials/PE-P1-svm-specA/cpp/trial_01/gen_00.cpp
// WL-SVM / PE-P1-svm-specA
// Soft-margin SVM trained by projected dual gradient ascent (Gauss-Seidel),
// implemented from the frozen specification using only general-purpose
// language facilities (no linear-algebra / tensor / statistics / BLAS facility).
//
// Logical modules: data, svm, app.

#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>

// ---------------------------------------------------------------------------
// frozen constants
// ---------------------------------------------------------------------------
namespace cfg {

constexpr int    D          = 4;
constexpr int    N_TRAIN_C1 = 100;
constexpr int    N_TRAIN_C2 = 100;
constexpr int    N_TEST_C1  = 50;
constexpr int    N_TEST_C2  = 50;
constexpr int    N          = N_TRAIN_C1 + N_TRAIN_C2;   // 200
constexpr int    N_TEST     = N_TEST_C1 + N_TEST_C2;     // 100

constexpr double MU1[D] = { 1.0, 1.0, 0.5, -0.5 };
constexpr double MU2[D] = { -1.0, -1.0, -0.5, 0.5 };

constexpr double SIGMA   = 0.8;
constexpr long long SEED = 1234567;

constexpr double C      = 10.0;
constexpr double LR     = 0.0001;
constexpr double LIMIT  = 0.0001;
constexpr int    SWEEPS = 1000;
constexpr double EPS_SV = 0.0000001;

}  // namespace cfg

// ---------------------------------------------------------------------------
// module: data  -- LCG-PM generator and dataset builder
// ---------------------------------------------------------------------------
namespace data {

// Lehmer / Park-Miller "minimal standard" multiplicative generator.
//   state = (48271 * state) mod 2147483647
class LcgPm {
 public:
  explicit LcgPm(long long seed) : state_(seed) {}

  long long next_state() {
    state_ = (MULT * state_) % MOD;   // exact in int64 (max ~1.037e14)
    return state_;
  }

  double next_uniform() {
    return static_cast<double>(next_state()) / 2147483647.0;
  }

  // Irwin-Hall(12) - 6, accumulated in ascending order.
  double next_normal() {
    double t = 0.0;
    for (int k = 0; k < 12; ++k) {
      t = t + next_uniform();
    }
    return t - 6.0;
  }

 private:
  static constexpr long long MULT = 48271;
  static constexpr long long MOD  = 2147483647;
  long long state_;
};

struct Dataset {
  // training set: class +1 block first, class -1 block second
  std::vector<std::vector<double>> x;   // N x D
  std::vector<double>              y;   // N
  // test set: class +1 block first, class -1 block second
  std::vector<std::vector<double>> xt;  // N_TEST x D
  std::vector<double>              yt;  // N_TEST
};

// Draw one block of `count` points with mean `mu`, points ascending,
// dimensions ascending.
static void draw_block(LcgPm& rng, int count, const double mu[cfg::D],
                       double label, std::vector<std::vector<double>>& xs,
                       std::vector<double>& ys) {
  for (int n = 0; n < count; ++n) {
    std::vector<double> p(cfg::D, 0.0);
    for (int d = 0; d < cfg::D; ++d) {
      p[d] = mu[d] + cfg::SIGMA * rng.next_normal();
    }
    xs.push_back(p);
    ys.push_back(label);
  }
}

Dataset build() {
  Dataset ds;
  LcgPm rng(cfg::SEED);

  // exact draw order, one stream, never reset
  draw_block(rng, cfg::N_TRAIN_C1, cfg::MU1, 1.0, ds.x, ds.y);    // 1
  draw_block(rng, cfg::N_TRAIN_C2, cfg::MU2, -1.0, ds.x, ds.y);   // 2
  draw_block(rng, cfg::N_TEST_C1, cfg::MU1, 1.0, ds.xt, ds.yt);   // 3
  draw_block(rng, cfg::N_TEST_C2, cfg::MU2, -1.0, ds.xt, ds.yt);  // 4

  return ds;
}

}  // namespace data

// ---------------------------------------------------------------------------
// module: svm  -- training, SV extraction, w/b recovery, f, g, evaluation
// ---------------------------------------------------------------------------
namespace svm {

struct Model {
  std::vector<double> alpha;
  double beta          = 1.0;
  bool   judge_last    = false;
  double error_last    = 0.0;
  double max_abs_delta = 0.0;

  std::vector<int> s_margin;
  std::vector<int> s_inside;
  std::vector<double> w;
  double b = 0.0;
};

// Mandatory precomputation of the Gram matrix.
std::vector<std::vector<double>> gram(const std::vector<std::vector<double>>& x) {
  const int n = static_cast<int>(x.size());
  std::vector<std::vector<double>> g(n, std::vector<double>(n, 0.0));
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
      double s = 0.0;
      for (int d = 0; d < cfg::D; ++d) {
        s = s + x[i][d] * x[j][d];
      }
      g[i][j] = s;
    }
  }
  return g;
}

Model train(const std::vector<std::vector<double>>& x,
            const std::vector<double>& y) {
  const int n = static_cast<int>(x.size());
  const std::vector<std::vector<double>> g = gram(x);

  Model m;
  m.alpha.assign(n, 0.0);
  m.beta = 1.0;

  bool   judge         = false;
  double error         = 0.0;
  double max_abs_delta = 0.0;

  for (int sweep = 0; sweep < cfg::SWEEPS; ++sweep) {
    judge         = false;
    error         = 0.0;
    max_abs_delta = 0.0;

    // (3.1) update alpha in place, ascending i (Gauss-Seidel)
    for (int i = 0; i < n; ++i) {
      const double yi = y[i];
      const std::vector<double>& gi = g[i];

      double item1 = 0.0;
      for (int j = 0; j < n; ++j) {
        item1 = item1 + m.alpha[j] * yi * y[j] * gi[j];
      }

      double item2 = 0.0;
      for (int j = 0; j < n; ++j) {
        item2 = item2 + m.alpha[j] * yi * y[j];
      }

      const double delta = 1.0 - item1 - m.beta * item2;
      const double ad    = delta < 0.0 ? -delta : delta;

      if (ad > max_abs_delta) max_abs_delta = ad;

      m.alpha[i] = m.alpha[i] + cfg::LR * delta;
      if (m.alpha[i] < 0.0) {
        m.alpha[i] = 0.0;
      } else if (m.alpha[i] > cfg::C) {
        m.alpha[i] = cfg::C;
      } else if (ad > cfg::LIMIT) {
        judge = true;
        error = error + (ad - cfg::LIMIT);
      }
    }

    // (3.2) update beta once per sweep
    double s = 0.0;
    for (int i = 0; i < n; ++i) {
      s = s + m.alpha[i] * y[i];
    }
    m.beta = m.beta + s * s / 2.0;
  }

  m.judge_last    = judge;
  m.error_last    = error;
  m.max_abs_delta = max_abs_delta;
  return m;
}

void recover(Model& m, const std::vector<std::vector<double>>& x,
             const std::vector<double>& y) {
  const int n = static_cast<int>(x.size());

  m.s_margin.clear();
  m.s_inside.clear();
  for (int i = 0; i < n; ++i) {
    if (cfg::EPS_SV < m.alpha[i] && m.alpha[i] < cfg::C - cfg::EPS_SV) {
      m.s_margin.push_back(i);
    }
  }
  for (int i = 0; i < n; ++i) {
    if (m.alpha[i] >= cfg::C - cfg::EPS_SV) {
      m.s_inside.push_back(i);
    }
  }

  m.w.assign(cfg::D, 0.0);
  for (int d = 0; d < cfg::D; ++d) {
    for (size_t k = 0; k < m.s_margin.size(); ++k) {
      const int i = m.s_margin[k];
      m.w[d] = m.w[d] + m.alpha[i] * y[i] * x[i][d];
    }
    for (size_t k = 0; k < m.s_inside.size(); ++k) {
      const int i = m.s_inside[k];
      m.w[d] = m.w[d] + m.alpha[i] * y[i] * x[i][d];
    }
  }

  double b = 0.0;
  for (size_t k = 0; k < m.s_margin.size(); ++k) {
    const int i = m.s_margin[k];
    double dp = 0.0;
    for (int d = 0; d < cfg::D; ++d) {
      dp = dp + m.w[d] * x[i][d];
    }
    b = b + (y[i] - dp);
  }
  if (m.s_margin.empty()) {
    std::fprintf(stderr, "FAIL: |S_margin| == 0\n");
    std::exit(1);
  }
  m.b = b / static_cast<double>(m.s_margin.size());
}

double f(const Model& m, const std::vector<double>& p) {
  double s = 0.0;
  for (int d = 0; d < cfg::D; ++d) {
    s = s + m.w[d] * p[d];
  }
  return s + m.b;
}

int g_sign(const Model& m, const std::vector<double>& p) {
  return f(m, p) >= 0.0 ? 1 : -1;
}

double objective(const std::vector<double>& alpha, const std::vector<double>& y,
                 const std::vector<std::vector<double>>& gm) {
  const int n = static_cast<int>(alpha.size());
  double sum_a = 0.0;
  for (int i = 0; i < n; ++i) {
    sum_a = sum_a + alpha[i];
  }
  double quad = 0.0;
  for (int i = 0; i < n; ++i) {
    double inner = 0.0;
    for (int j = 0; j < n; ++j) {
      inner = inner + alpha[i] * alpha[j] * y[i] * y[j] * gm[i][j];
    }
    quad = quad + inner;
  }
  return sum_a - 0.5 * quad;
}

double alpha_checksum(const std::vector<double>& alpha) {
  const int n = static_cast<int>(alpha.size());
  double s = 0.0;
  for (int i = 0; i < n; ++i) {
    s = s + alpha[i] * static_cast<double>((i % 97) + 1);
  }
  return s;
}

}  // namespace svm

// ---------------------------------------------------------------------------
// module: app  -- entry point and output formatting
// ---------------------------------------------------------------------------
namespace app {

// Fixed-point formatting with the negative-zero normalization required by 1.4.
std::string fixed(double v, int digits) {
  if (v == 0.0) v = 0.0;  // maps -0.0 to +0.0
  char buf[64];
  std::snprintf(buf, sizeof(buf), "%.*f", digits, v);
  std::string s(buf);
  // guard against "-0.000000" produced by rounding a tiny negative value
  bool all_zero = true;
  for (size_t i = 0; i < s.size(); ++i) {
    const char c = s[i];
    if (c >= '1' && c <= '9') { all_zero = false; break; }
  }
  if (all_zero && !s.empty() && s[0] == '-') s.erase(s.begin());
  return s;
}

std::string f6(double v)  { return fixed(v, 6); }
std::string f12(double v) { return fixed(v, 12); }

int run() {
  const data::Dataset ds = data::build();

  svm::Model m = svm::train(ds.x, ds.y);
  svm::recover(m, ds.x, ds.y);

  const std::vector<std::vector<double>> gm = svm::gram(ds.x);

  // evaluation
  int train_correct = 0;
  for (int i = 0; i < cfg::N; ++i) {
    const int pred = svm::g_sign(m, ds.x[i]);
    if (static_cast<double>(pred) == ds.y[i]) ++train_correct;
  }

  std::vector<int> pred(cfg::N_TEST, 0);
  int correct_c1 = 0, correct_c2 = 0;
  for (int i = 0; i < cfg::N_TEST; ++i) {
    pred[i] = svm::g_sign(m, ds.xt[i]);
    if (static_cast<double>(pred[i]) == ds.yt[i]) {
      if (i < cfg::N_TEST_C1) ++correct_c1; else ++correct_c2;
    }
  }
  const int test_correct = correct_c1 + correct_c2;

  double alpha_sum = 0.0;
  for (int i = 0; i < cfg::N; ++i) alpha_sum = alpha_sum + m.alpha[i];
  double alpha_y_sum = 0.0;
  for (int i = 0; i < cfg::N; ++i) alpha_y_sum = alpha_y_sum + m.alpha[i] * ds.y[i];

  const double obj  = svm::objective(m.alpha, ds.y, gm);
  const double chks = svm::alpha_checksum(m.alpha);

  const double train_acc = static_cast<double>(train_correct) / static_cast<double>(cfg::N);
  const double test_acc  = static_cast<double>(test_correct) / static_cast<double>(cfg::N_TEST);
  const double test_acc_c1 = static_cast<double>(correct_c1) / static_cast<double>(cfg::N_TEST_C1);
  const double test_acc_c2 = static_cast<double>(correct_c2) / static_cast<double>(cfg::N_TEST_C2);

  std::string out;
  out += "SVM_VERSION 1\n";
  out += "SWEEPS " + std::to_string(cfg::SWEEPS) + "\n";
  out += "CONVERGED " + std::string(m.judge_last ? "0" : "1") + "\n";
  out += "ERROR_LAST " + f6(m.error_last) + "\n";
  out += "MAX_ABS_DELTA " + f6(m.max_abs_delta) + "\n";
  out += "BETA " + f6(m.beta) + "\n";
  out += "NS_MARGIN " + std::to_string(m.s_margin.size()) + "\n";
  out += "NS_INSIDE " + std::to_string(m.s_inside.size()) + "\n";
  out += "W";
  for (int d = 0; d < cfg::D; ++d) out += " " + f6(m.w[d]);
  out += "\n";
  out += "B " + f6(m.b) + "\n";
  out += "OBJECTIVE " + f6(obj) + "\n";
  out += "ALPHA_SUM " + f6(alpha_sum) + "\n";
  out += "ALPHA_Y_SUM " + f12(alpha_y_sum) + "\n";
  out += "ALPHA_CHECKSUM " + f6(chks) + "\n";
  out += "TRAIN_ACC " + f6(train_acc) + "\n";
  out += "TEST_ACC " + f6(test_acc) + "\n";
  out += "TEST_ACC_C1 " + f6(test_acc_c1) + "\n";
  out += "TEST_ACC_C2 " + f6(test_acc_c2) + "\n";
  out += "TEST_CORRECT " + std::to_string(correct_c1) + " " +
         std::to_string(correct_c2) + " " + std::to_string(test_correct) + "\n";
  out += "PRED";
  for (int i = 0; i < cfg::N_TEST; ++i) out += " " + std::to_string(pred[i]);
  out += "\n";
  out += "ALPHA";
  for (int i = 0; i < cfg::N; ++i) out += " " + f6(m.alpha[i]);
  out += "\n";

  std::fwrite(out.data(), 1, out.size(), stdout);
  return 0;
}

}  // namespace app

int main() {
  return app::run();
}

File: benchmark/2026-09-17-7677581(latest)/llm-practical/trials/PE-P1-svm-specA/swift/trial_01/gen_00.swift
// WL-SVM / PE-P1-svm-specA -- Swift implementation
// Standard library only. No linear-algebra / tensor / statistics / BLAS facility used.
// Single threaded, no SIMD, no intrinsics.

import Foundation

// ===========================================================================
// MARK: - Frozen constants
// ===========================================================================

enum Const {
    static let D = 4
    static let N_TRAIN_C1 = 100
    static let N_TRAIN_C2 = 100
    static let N_TEST_C1 = 50
    static let N_TEST_C2 = 50
    static let MU1: [Double] = [1.0, 1.0, 0.5, -0.5]
    static let MU2: [Double] = [-1.0, -1.0, -0.5, 0.5]
    static let SIGMA = 0.8
    static let SEED = 1234567
    static let C = 10.0
    static let LR = 0.0001
    static let LIMIT = 0.0001
    static let SWEEPS = 1000
    static let EPS_SV = 0.0000001
}

// ===========================================================================
// MARK: - module `data` : LCG-PM generator and dataset builder
// ===========================================================================

enum Data {

    /// Lehmer / Park-Miller minimal standard multiplicative generator.
    /// state_{n+1} = (48271 * state_n) mod 2147483647
    struct LCGPM {
        static let MULT = 48271
        static let MOD = 2147483647
        private var state: Int

        init(seed: Int) {
            self.state = seed
        }

        mutating func nextState() -> Int {
            state = (LCGPM.MULT * state) % LCGPM.MOD
            return state
        }

        mutating func nextUniform() -> Double {
            return Double(nextState()) / 2147483647.0
        }

        /// Irwin-Hall(12) - 6, ascending accumulation, never reassociated.
        mutating func nextNormal() -> Double {
            var t = 0.0
            for _ in 0..<12 {
                t = t + nextUniform()
            }
            return t - 6.0
        }
    }

    /// The complete dataset, flattened row-major (`D` doubles per point).
    struct Dataset {
        var xTrain: [Double]   // (N_TRAIN_C1 + N_TRAIN_C2) * D
        var yTrain: [Double]   // +1 for i < N_TRAIN_C1, -1 afterwards
        var xTestC1: [Double]  // N_TEST_C1 * D
        var xTestC2: [Double]  // N_TEST_C2 * D
        var n: Int
    }

    /// Draws one block of `count` points around `mu`, points ascending,
    /// dimensions ascending, from the single shared stream.
    private static func drawBlock(_ rng: inout LCGPM, _ mu: [Double], _ count: Int) -> [Double] {
        var out = [Double](repeating: 0.0, count: count * Const.D)
        for n in 0..<count {
            for d in 0..<Const.D {
                out[n * Const.D + d] = mu[d] + Const.SIGMA * rng.nextNormal()
            }
        }
        return out
    }

    /// Exact draw order of the specification: train +1, train -1, test +1, test -1.
    static func build() -> Dataset {
        var rng = LCGPM(seed: Const.SEED)

        let trainC1 = drawBlock(&rng, Const.MU1, Const.N_TRAIN_C1)
        let trainC2 = drawBlock(&rng, Const.MU2, Const.N_TRAIN_C2)
        let testC1 = drawBlock(&rng, Const.MU1, Const.N_TEST_C1)
        let testC2 = drawBlock(&rng, Const.MU2, Const.N_TEST_C2)

        // Concatenate class +1 first, class -1 second.
        var xTrain = [Double]()
        xTrain.reserveCapacity(trainC1.count + trainC2.count)
        xTrain.append(contentsOf: trainC1)
        xTrain.append(contentsOf: trainC2)

        var yTrain = [Double](repeating: 1.0, count: Const.N_TRAIN_C1 + Const.N_TRAIN_C2)
        for i in Const.N_TRAIN_C1..<(Const.N_TRAIN_C1 + Const.N_TRAIN_C2) {
            yTrain[i] = -1.0
        }

        return Dataset(xTrain: xTrain,
                       yTrain: yTrain,
                       xTestC1: testC1,
                       xTestC2: testC2,
                       n: Const.N_TRAIN_C1 + Const.N_TRAIN_C2)
    }
}

// ===========================================================================
// MARK: - module `svm` : training, SV extraction, w/b recovery, f, g, eval
// ===========================================================================

enum SVM {

    struct TrainResult {
        var alpha: [Double]
        var beta: Double
        var judge: Bool
        var errorLast: Double
        var maxAbsDelta: Double
        var gram: [Double]
    }

    struct Model {
        var w: [Double]
        var b: Double
        var nsMargin: Int
        var nsInside: Int
    }

    /// Mandatory one-off Gram build: G[i][j] = sum_d x[i][d] * x[j][d].
    static func buildGram(_ x: [Double], _ n: Int) -> [Double] {
        var g = [Double](repeating: 0.0, count: n * n)
        for i in 0..<n {
            for j in 0..<n {
                var s = 0.0
                for d in 0..<Const.D {
                    s = s + x[i * Const.D + d] * x[j * Const.D + d]
                }
                g[i * n + j] = s
            }
        }
        return g
    }

    /// Exactly SWEEPS sweeps of in-place (Gauss-Seidel) dual ascent, no early exit.
    static func train(_ x: [Double], _ y: [Double], _ n: Int) -> TrainResult {
        let g = buildGram(x, n)

        var alpha = [Double](repeating: 0.0, count: n)
        var beta = 1.0

        var judge = false
        var error = 0.0
        var maxAbsDelta = 0.0

        for _ in 0..<Const.SWEEPS {

            judge = false
            error = 0.0
            maxAbsDelta = 0.0

            // (3.1) update alpha, ascending i, in place
            for i in 0..<n {
                let yi = y[i]
                let rowBase = i * n

                var item1 = 0.0
                for j in 0..<n {
                    item1 = item1 + alpha[j] * yi * y[j] * g[rowBase + j]
                }

                var item2 = 0.0
                for j in 0..<n {
                    item2 = item2 + alpha[j] * yi * y[j]
                }

                let delta = 1.0 - item1 - beta * item2
                let ad = abs(delta)

                if ad > maxAbsDelta { maxAbsDelta = ad }

                alpha[i] = alpha[i] + Const.LR * delta
                if alpha[i] < 0.0 {
                    alpha[i] = 0.0
                } else if alpha[i] > Const.C {
                    alpha[i] = Const.C
                } else if ad > Const.LIMIT {
                    judge = true
                    error = error + (ad - Const.LIMIT)
                }
            }

            // (3.2) update beta, once per sweep
            var s = 0.0
            for i in 0..<n {
                s = s + alpha[i] * y[i]
            }
            beta = beta + s * s / 2.0
        }

        return TrainResult(alpha: alpha,
                           beta: beta,
                           judge: judge,
                           errorLast: error,
                           maxAbsDelta: maxAbsDelta,
                           gram: g)
    }

    /// Support-vector extraction plus w / b recovery.
    static func recover(_ alpha: [Double], _ x: [Double], _ y: [Double], _ n: Int) -> Model {
        var sMargin = [Int]()
        var sInside = [Int]()
        for i in 0..<n {
            if Const.EPS_SV < alpha[i] && alpha[i] < Const.C - Const.EPS_SV {
                sMargin.append(i)
            }
            if alpha[i] >= Const.C - Const.EPS_SV {
                sInside.append(i)
            }
        }

        var w = [Double](repeating: 0.0, count: Const.D)
        for d in 0..<Const.D {
            for i in sMargin {
                w[d] = w[d] + alpha[i] * y[i] * x[i * Const.D + d]
            }
            for i in sInside {
                w[d] = w[d] + alpha[i] * y[i] * x[i * Const.D + d]
            }
        }

        var b = 0.0
        for i in sMargin {
            var dp = 0.0
            for d in 0..<Const.D {
                dp = dp + w[d] * x[i * Const.D + d]
            }
            b = b + (y[i] - dp)
        }
        b = b / Double(sMargin.count)

        return Model(w: w, b: b, nsMargin: sMargin.count, nsInside: sInside.count)
    }

    static func f(_ model: Model, _ p: [Double], _ off: Int) -> Double {
        var s = 0.0
        for d in 0..<Const.D {
            s = s + model.w[d] * p[off + d]
        }
        return s + model.b
    }

    static func g(_ model: Model, _ p: [Double], _ off: Int) -> Int {
        return f(model, p, off) >= 0.0 ? 1 : -1
    }

    /// Counts correct predictions over a block whose true label is `label`.
    static func countCorrect(_ model: Model, _ block: [Double], _ count: Int, _ label: Int) -> Int {
        var c = 0
        for i in 0..<count {
            if g(model, block, i * Const.D) == label { c += 1 }
        }
        return c
    }

    static func objective(_ alpha: [Double], _ y: [Double], _ gram: [Double], _ n: Int) -> Double {
        var sumAlpha = 0.0
        for i in 0..<n {
            sumAlpha = sumAlpha + alpha[i]
        }
        var outer = 0.0
        for i in 0..<n {
            var inner = 0.0
            for j in 0..<n {
                inner = inner + alpha[i] * alpha[j] * y[i] * y[j] * gram[i * n + j]
            }
            outer = outer + inner
        }
        return sumAlpha - 0.5 * outer
    }
}

// ===========================================================================
// MARK: - module `app` : entry point and output formatting
// ===========================================================================

enum App {

    /// Fixed-point, exactly 6 fractional digits, negative zero normalized away.
    static func f6(_ v: Double) -> String {
        let x = (v == 0.0) ? 0.0 : v
        return String(format: "%.6f", x)
    }

    /// Fixed-point, exactly 12 fractional digits, negative zero normalized away.
    static func f12(_ v: Double) -> String {
        let x = (v == 0.0) ? 0.0 : v
        return String(format: "%.12f", x)
    }

    static func run() {
        let ds = Data.build()
        let n = ds.n

        let tr = SVM.train(ds.xTrain, ds.yTrain, n)
        let model = SVM.recover(tr.alpha, ds.xTrain, ds.yTrain, n)

        // Dual sums
        var alphaSum = 0.0
        for i in 0..<n { alphaSum = alphaSum + tr.alpha[i] }

        var alphaYSum = 0.0
        for i in 0..<n { alphaYSum = alphaYSum + tr.alpha[i] * ds.yTrain[i] }

        var alphaChecksum = 0.0
        for i in 0..<n {
            alphaChecksum = alphaChecksum + tr.alpha[i] * Double((i % 97) + 1)
        }

        let obj = SVM.objective(tr.alpha, ds.yTrain, tr.gram, n)

        // Training accuracy over the 200 training points, +1 block then -1 block.
        var trainCorrect = 0
        for i in 0..<n {
            let pred = SVM.g(model, ds.xTrain, i * Const.D)
            let truth = ds.yTrain[i] > 0.0 ? 1 : -1
            if pred == truth { trainCorrect += 1 }
        }
        let trainAcc = Double(trainCorrect) / Double(n)

        // Test predictions and accuracies.
        var preds = [Int]()
        preds.reserveCapacity(Const.N_TEST_C1 + Const.N_TEST_C2)
        for i in 0..<Const.N_TEST_C1 {
            preds.append(SVM.g(model, ds.xTestC1, i * Const.D))
        }
        for i in 0..<Const.N_TEST_C2 {
            preds.append(SVM.g(model, ds.xTestC2, i * Const.D))
        }

        let correctC1 = SVM.countCorrect(model, ds.xTestC1, Const.N_TEST_C1, 1)
        let correctC2 = SVM.countCorrect(model, ds.xTestC2, Const.N_TEST_C2, -1)
        let correctTotal = correctC1 + correctC2

        let testAcc = Double(correctTotal) / Double(Const.N_TEST_C1 + Const.N_TEST_C2)
        let testAccC1 = Double(correctC1) / Double(Const.N_TEST_C1)
        let testAccC2 = Double(correctC2) / Double(Const.N_TEST_C2)

        var out = ""
        out += "SVM_VERSION 1\n"
        out += "SWEEPS \(Const.SWEEPS)\n"
        out += "CONVERGED \(tr.judge ? 0 : 1)\n"
        out += "ERROR_LAST \(f6(tr.errorLast))\n"
        out += "MAX_ABS_DELTA \(f6(tr.maxAbsDelta))\n"
        out += "BETA \(f6(tr.beta))\n"
        out += "NS_MARGIN \(model.nsMargin)\n"
        out += "NS_INSIDE \(model.nsInside)\n"
        out += "W " + model.w.map { f6($0) }.joined(separator: " ") + "\n"
        out += "B \(f6(model.b))\n"
        out += "OBJECTIVE \(f6(obj))\n"
        out += "ALPHA_SUM \(f6(alphaSum))\n"
        out += "ALPHA_Y_SUM \(f12(alphaYSum))\n"
        out += "ALPHA_CHECKSUM \(f6(alphaChecksum))\n"
        out += "TRAIN_ACC \(f6(trainAcc))\n"
        out += "TEST_ACC \(f6(testAcc))\n"
        out += "TEST_ACC_C1 \(f6(testAccC1))\n"
        out += "TEST_ACC_C2 \(f6(testAccC2))\n"
        out += "TEST_CORRECT \(correctC1) \(correctC2) \(correctTotal)\n"
        out += "PRED " + preds.map { String($0) }.joined(separator: " ") + "\n"
        out += "ALPHA " + tr.alpha.map { f6($0) }.joined(separator: " ") + "\n"

        print(out, terminator: "")
    }
}

App.run()
